Mathlib/Data/Nat/Fib/Lemmas.lean

```diff
@@ -0,0 +1,79 @@
+/-
+Copyright (c) 2025 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import Mathlib.Data.Nat.Fib.Basic
+import Mathlib.LinearAlgebra.Matrix.Notation
+import Mathlib.LinearAlgebra.Matrix.Determinant.Basic
+
+/-!
+# Cassini and Catalan identities for the Fibonacci numbers
+
+Cassini's identity states that for nonzero `n`, `fib (n + 1) * fib (n - 1) - fib n ^ 2` is equal
+to `(-1) ^ n`. And Catalan's identity states that for nonzero `x` and `a`, we get
+`fib (x + a) ^ 2 - fib x * fib (x + 2 * a) = (-1) ^ x * fib a ^ 2`.
+-/
+
+namespace Nat
+
+/-- For any nonzero `n`, the matrix `[fib (n + 1), fib n; fib n, fib (n - 1)]` is equal to
+`[1, 1; 1, 0] ^ n`. -/
+lemma fib_matrix_eq : ∀ {n : ℕ}, n ≠ 0 →
+    !![fib (n + 1), fib n; fib n, fib (n - 1)] = !![1, 1; 1, 0] ^ n
+  | 1, _ => by simp
+  | n + 2, _ => by
+    rw [pow_succ, ← fib_matrix_eq (add_one_ne_zero n)]
+    simp [fib_add_two, ← add_assoc, add_rotate, add_comm (fib n)]
+
+/-- **Cassini's identity**: for nonzero `n`, `fib (n + 1) * fib (n - 1) - fib n ^ 2 = (-1) ^ n`. -/
+lemma fib_succ_mul_fib_pred_sub_fib_sq {n : ℕ} (hn : n ≠ 0) :
+    (fib (n + 1) * fib (n - 1) - fib n ^ 2 : ℤ) = (-1) ^ n :=
+  calc _ = (!![fib (n + 1), fib n; fib n, fib (n - 1)] : Matrix _ _ ℤ).det := by simp [pow_two]
+    _ = (!![fib (n + 1), fib n; fib n, fib (n - 1)].map (castRingHom ℤ)).det := rfl
+    _ = (!![1, 1; 1, 0].map (castRingHom ℤ) ^ n).det := by rw [fib_matrix_eq hn, Matrix.map_pow]
+    _ = (!![1, 1; 1, 0] ^ n).det := by congr; simp [← Matrix.ext_iff]
+    _ = (-1) ^ n := by simp
+
+/-- Cassini's identity for nonzero even `n`: `fib (n + 1) * fib (n - 1) = fib n ^ 2 + 1.` -/
+lemma fib_succ_mul_fib_pred_of_ne_zero_even {n : ℕ} (hn : n ≠ 0) (hn2 : Even n) :
+    fib (n + 1) * fib (n - 1) = fib n ^ 2 + 1 := by
+  obtain ⟨a, rfl⟩ := hn2
+  grind [fib_succ_mul_fib_pred_sub_fib_sq, pow_mul]
+
+/-- Cassini's identity for odd `n`: `fib (n + 1) * fib (n - 1) = fib n ^ 2 - 1.` -/
+lemma fib_succ_mul_fib_pred_of_odd {n : ℕ} (hn : Odd n) :
+    fib (n + 1) * fib (n - 1) = fib n ^ 2 - 1 := by
+  obtain ⟨a, rfl⟩ := hn
+  grind [fib_succ_mul_fib_pred_sub_fib_sq, pow_mul]
+
+/-- **Catalan's identity**: for nonzero `x` and `a`,
+`fib (x + a) ^ 2 - fib x * fib (x + 2 * a) = (-1) ^ x * fib a ^ 2`. -/
+lemma fib_add_sq_sub_fib_mul_fib_add_two_mul {x a : ℕ} (hx : x ≠ 0) (ha : a ≠ 0) :
```

Comment:
Cassini is just the special case `a = 1`, right? Can you reprove it as a corollary of this result?